[Onnx] Add constant folding to ONNX graph during export (Resubmission)

spandantiwari opened the pull request:
Rewritten version of https://github.com/pytorch/pytorch/pull/17771 using graph C++ APIs. 

This PR adds the ability to do constant folding on ONNX graphs during PT->ONNX export. This is done mainly to optimize the graph and make it leaner. The two attached snapshots show a multiple-node LSTM model before and after constant folding. 
A couple of notes:
1. Constant folding is by default turned off for now. The goal is to turn it on by default once we have validated it through all the tests.
2. Support for folding in nested blocks is not in place, but will be added in the future, if needed. 

**Original Model:**
![multiple_lstm_original](https://user-images.githubusercontent.com/23646532/53987630-6ac53980-40d6-11e9-9702-1ccfee124a83.JPG)
**Constant-folded model:**
![multiple_lstm_constant_folded](https://user-images.githubusercontent.com/23646532/53987632-6c8efd00-40d6-11e9-81c5-362c16f68861.JPG)

Commit:
Add constant folding optimization to ONNX graph export.

Commit:
Removed print statement from entrypoint in init.cpp.

spandantiwari commented:
I see a failing test in one of the runs, but it does not seem related to this change. Also, I cannot repro it locally. @suo - Could you please help take a quick look?

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -155,7 +155,7 @@ def run_debug_test(self, model, train, batch_size, state_dict=None,
     def run_actual_test(self, model, train, batch_size, state_dict=None,
                         input=None, use_gpu=True, rtol=0.001, atol=1e-7,
-                        example_outputs=None):
+                        example_outputs=None, do_constant_folding=False):
```
I am wondering if we enable the flag, how many tests will pass?

@spandantiwari do you want to give it a try?

spandantiwari replied:
@houseroad, maybe as the next PR. :-)

houseroad replied:
i mean locally first, i am wondering how stable it is :-)

houseroad commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+}
+
+} // namespace jit
+} // namespace torch
```
new line character

spandantiwari replied:
Fixed.

houseroad commented on torch/onnx/utils.py:
```diff
@@ -257,6 +257,9 @@ def _model_to_graph(model, args, f, verbose=False, training=False,
     params_dict = dict(zip(param_names, params))
 
+    if do_constant_folding:
+        params_dict = torch._C._jit_pass_onnx_constant_fold(graph, params_dict)
```
yeah, we have to do this pass after we get the param_dict.

spandantiwari replied:
Yes.

houseroad reviewed (commented):
briefly scanned the pr. This is really useful. It makes the onnx graph much simpler (and maybe much faster). Thanks!

Will do a thorough pass tomorrow or weekend.

houseroad commented:
btw, the failing test should be unrelaed, i have restarted it

suo reviewed (changes requested):
Looks like a great start! I have only reviewed the constant folding pass so far (not yet at the `runTorchBackendForOnnx()` or testing stuff).

The pass looks generally correct, so my comments are mostly about making things more robust/idiomatic with the rest of the JIT code. I am going to put some general themes here and refer to them inline.

1. On `uniqueName()`: we should avoid using `uniqueName()` to identify a value. This is fragile because if you ever call setUniqueName("foo"), any value that previously had the name "foo" will get renamed something else (like "foo.1". So values can get arbitrarily renamed by graph transformations, making the name an unreliable way to reference any given value.

We typically use the direct `Value*` to reference a value. It's guaranteed to be a unique *and* stable identifier.

2. On paramDict: This uses `uniqueName()`, which is generally an anti-pattern. However, because the onnx export pass manually assigns names to the parameter values, we can be sure that at the beginning of the pass at least, the names refer to the right values in the graph.

So in keeping with using `Value*` to reference values (as mentioned in [1]), I think the first thing the pass should do is map that input `Value*` to its corresponding (name, tensor), and use that mapping as a source of truth. So isParam becomes `isParam(Value* v)`, which just checks whether v is in the `std::unordered_map<Value*, std::pair<std::string, at::Tensor>>` (feel free to make that data structure whatever you prefer). 

At the end of the pass, you can use that same mapping to fix up `paramDict` to erase the folded parameters.

2. It seems like the fact that constants could be `onnx::constant` *or* a parameter value is adding a lot of complexity to the pass. We're doing a lot of tracking for which input is which constant kind.

If you had the data structure I mentioned in [2], `isParam()` is an easy map lookup and thus you don't have to "save" that info when checking node inputs for constants, which eliminates the need for `kindOfLeafNode` and its cousin, `ConstantLeafNodeKind`.

Commit:
Refactoring - initial checkpoint.

Commit:
Refactored to move away from name-based logic.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+// This method updates the block in-place to fold all the one-time 
+// constant-based computations/ops into an initializer node.
+void ConstantFoldONNX(Block* b, std::map<std::string, at::Tensor>& paramsDict) {
+  auto sourceNode = getSourceNode(*b);
```
That's useful. Thanks for pointing to it.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+using namespace ::c10::onnx;
+}
+
+static Node* getSourceNode(Block& b) {
```
Fixed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+// constant-based computations/ops into an initializer node.
+void ConstantFoldONNX(Block* b, std::map<std::string, at::Tensor>& paramsDict) {
+  auto sourceNode = getSourceNode(*b);
+  if (sourceNode == nullptr) {
```
Done.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+  // not supported for now.
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    auto node = *it;
+    auto nodeKind = node->kind().toDisplayString();
```
Removed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+        AT_ASSERT(sourceNode == inputNode); // One and only one prim::Param node in block.
+      }
+      bool isConstant = inputNode->kind() == onnx::Constant && !inputNode->mustBeNone()
+        && toString(inputNode->kindOfS("value")) == std::string("t"); // TODO: Check other types?
```
Yes. Fixed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+      // Disconnect the folded node. Create a new initializer for the folded  
+      // tensor and replace the output of the folded node with the 
+      // initializer as input for all downstream nodes.
+      auto newSourceNodeOutput = sourceNode->addOutput();
```
As we discussed, the use of initializers is preferred from ONNX point of view (to distinguish between learnable parameters and real constants.), hence keeping it.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+      // Find the indices to outputs of the source node that are
+      // feeding into the folded node. Used below for removing 
+      // corresponding entried in params_dict.
+      std::unordered_map<std::string, size_t> sourceOutputNames;
```
Fixed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+        sourceOutputNames[sourceNode->outputs().at(i)->uniqueName()] = i;
+      }
+      for (size_t i = 0; i < numInputs; ++i) { 
+        if (kindOfLeafNode[i] == ConstantLeafNodeKind::PRIM_PARAM) {
```
Agreed. Fixed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+      }
+      node->removeAllInputs();
+      for (const auto& elem : sourceOutputsToRemove) {
+        if (!sourceNode->outputs().at(elem.first)->hasUses()) {
```
Done.

spandantiwari commented:
@suo - thanks for your thoughtful review. I have revised the PR extensively based on your feedback. Could you please take a look and see if it looks good.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+        kindOfLeafNode.push_back(ConstantLeafNodeKind::PRIM_PARAM);
+      }
+      else if (isConstant) {
+        inputTensorValues.push_back(inputNode->t(c10::Symbol::fromQualString("attr::value")));
```
Fixed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+        // Constant folding not supported for this op. Skip it.
+        continue;
+      }
+      if (node->outputs().size() > 1) {
```
Good suggestion. Done.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,192 @@
+    inputTensorValues.reserve(numInputs);
+    std::vector<ConstantLeafNodeKind> kindOfLeafNode;
+    kindOfLeafNode.reserve(numInputs);
+    for (auto val : node->inputs()) {
```
Done.

Commit:
Rebasing to latest master. Fixing conflict, and minor lint changes.

Commit:
Fixing passing of do_constant_folding argument to model_to_graph.

spandantiwari commented:
@houseroad, @suo - a lot of the failures, maybe most of them, are unrelated.

Commit:
Reverting test expect file.

spandantiwari commented:
Doing another run.

houseroad commented:
@spandantiwari circle ci is down now, let's wait a bit and restart all the tests.

spandantiwari commented:
@houseroad - OK. Thanks for letting me know.

suo commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+// This method updates the block in-place to fold all the one-time 
+// constant-based computations/ops into an initializer node.
+void ConstantFoldONNX(Block* b, ParamMap& paramsDict) {
+  auto sourceNode = b->param_node();
```
These changes look good! My one remaining nit is that we shouldn't operate on an entity called `sourceNode`, since it more clearly expresses your intent to operate on block inputs (as a reminder, `block->inputs()` is equivalent to `param_node()->outputs()`).

spandantiwari replied:
Good point. Updated.

suo commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+}
+
+static at::Tensor runTorchBackendForOnnx(const Node* node, std::vector<at::Tensor>& inputTensorValues) {
+  at::Tensor updated_val;
```
@houseroad, can you take a look at these cases? I'm not familiar with the onnx->aten semantics

spandantiwari replied:
Yes, that will be great. Just fyi - I have tested this part on multiple large production models, including BERT, to make sure that these computations are correct, and the numbers match exactly.

spandantiwari replied:
Updated based on @houseroad 's feedback on this section.

suo commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+    updated_val = inputTensorValues[0].permute(node->is(attr::perm));
+  }
+  else {
+    updated_val = at::empty({0});
```
I think it would be clearer for you to return a c10::optional<at::Tensor> from this function, and return `c10::nullopt` in the case that `node` is not a supported op for constant folding.

spandantiwari replied:
Done.

suo commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+        continue;
+    }
+    auto updated_val = runTorchBackendForOnnx(node, inputTensorValues);
+    if (updated_val.size(0) == 0) {
```
same as above; returning `c10::nullopt` instead of an empty tensor will express your intent more clearly.

spandantiwari replied:
Done.

suo reviewed (commented):
Looks pretty good from my side! Not sure about the onnx->aten implementations, so I've asked @houseroad to eyeball those.

houseroad commented:
Import to check whether it breaks anything internal :-)

houseroad commented on torch/onnx/utils.py:
```diff
@@ -275,6 +275,9 @@ def _model_to_graph(model, args, f, verbose=False, training=False,
     params_dict = dict(zip(param_names, params))
 
+    if do_constant_folding:
+        params_dict = torch._C._jit_pass_onnx_constant_fold(graph, params_dict)
```
We can also check whether the opset is supported by our constant folding. Op spec may change, and the optimization may not be applicable any more.

spandantiwari replied:
Yes, good point. Now checking for opset version (opset 9).

houseroad commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+        // Constant folding for multiple-output nodes not supported. Skip it.
+        continue;
+      }
+    if (!std::all_of(node->inputs().begin(), node->inputs().end(),
```
can we extract this as a helper function?

spandantiwari replied:
OK. Done.

houseroad commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+using ParamMap = std::map<std::string, at::Tensor>;
+using ValueToParamPairMap = std::map<Value*, std::pair<std::string, at::Tensor>>;
+
+static ValueToParamPairMap buildValueToParamsMap(Block* b, const ParamMap& paramsDict) {
```
We can put everything into anonymous namespace except `ConstantFoldONNX`. So they don't need to be static i think.

spandantiwari replied:
I think both are viable and have the same effect. But your suggestion is cleaner. Updated with anonymous namespace.

houseroad reviewed (commented):
Great work. This is very helpful. For testing, we should create test cases to cover 1) Slice, 2) Concat, 3) Transpose, 4) Unsqueeze, 5) No optimization.

It can be added here: https://github.com/pytorch/pytorch/blob/master/test/onnx/test_utility_funs.py
We can just run the optimization pass, and iterate over the graph to check the change after the optimization. Here are some examples how we check the optimizer work in ONNX: https://github.com/onnx/onnx/blob/master/onnx/test/optimizer_test.py#L137

Commit:
Removing direct use of prim::Param node. Using c10::optional.

Commit:
Misc updates based on another round of feedback.

Commit:
Another lint remooval.

Commit:
Adding unit tests for constant folding.

spandantiwari commented on test/onnx/verify.py:
```diff
@@ -434,6 +437,8 @@ def run_helper(torch_out, args):
             if isinstance(torch_out, torch.Tensor):
                 torch_out = (torch_out,)
+            if any(isinstance(o, tuple) for o in torch_out):  # LSTM and GRU outputs can have nested tuples
+                torch_out, _ = torch._C._jit_flatten(torch_out)
```
I tried all the tests in test_pytorch_onnx_caffe2.py and nothing seems to break when we call _jit_flatten unconditionally. Conditional removed.

spandantiwari commented on torch/onnx/utils.py:
```diff
@@ -334,7 +339,7 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
                                                         output_names, operator_export_type,
                                                         example_outputs, propagate,
-                                                        _retain_param_name)
+                                                        _retain_param_name, do_constant_folding)
```
Done.

spandantiwari commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -176,7 +176,8 @@ def run_actual_test(self, model, train, batch_size, state_dict=None,
+        verify.verify(model, input, c2, rtol=rtol, atol=atol,
+                      do_constant_folding=do_constant_folding)
 
     def run_model_test(self, model, train, batch_size, state_dict=None,
```
Yes, sure. Updated.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+
+#if defined(_MSC_VER)
+#include <BaseTsd.h>
+typedef SSIZE_T ssize_t;
```
We don't. Removed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+}
+
+static void eraseUnusedNodeOutputs(Node* node) {
+  std::vector<std::string> removedOutputNames;
```
Vestige from older version. Removed.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
```
This is so that the ops are identified with small_ish_ but still qualified names such as onnx::Slice etc. Mostly to improve code readability. Also other passes follow similar style.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+    auto startsAttr = node->is(attr::starts);
+    auto endsAttr = node->is(attr::ends);
+    if (axesAttr.size() != startsAttr.size() || axesAttr.size() != endsAttr.size()) {
+      throw std::runtime_error("onnx::Slice node attributues named, axes, starts, and ends, must be the same length.");
```
OK. Removed the exception, just skipping node instead.

spandantiwari commented on torch/csrc/jit/passes/onnx/constant_fold.cpp:
```diff
@@ -0,0 +1,187 @@
+    }
+  }  
+  else if (node->kind() == onnx::Concat) {
+    updated_val = at::cat(at::TensorList(inputTensorValues), node->i(attr::axis));
```
Done.

spandantiwari commented:
@houseroad and @suo - Thanks for your feedbacks. I have updated the PR. Could you please review and see if this is good for merge.

spandantiwari commented:
> Great work. This is very helpful. For testing, we should create test cases to cover 1) Slice, 2) Concat, 3) Transpose, 4) Unsqueeze, 5) No optimization.
> 
@houseroad - As per your suggestion, I have added tests in `test_utility_funs.py` that test constant folding at the graph level. Between the two test points that I have added, I am testing all the nodes that are currently supported in onnx constant folding pass. 

I could not add test for individual ops because if we create a minimal model with a given op, then the `torch._C._jit_pass_constant_propagation` pass jit pass optimizes it at the jit graph level, and it does not show up in the onnx graph at all. Therefore, we cannot write a viable test using those models. The constant folding pass is aimed at folding those nodes that are added during one of the onnx passes. Therefore, in our tests we are limited to using those ops that produce constant-foldable nodes during the onnx passes. LSTM/GRU ops is one such example. That is why I have used these ops in my test points. Additionally, the test points are checking the end-to-end model also, as you had recommended. 

Between the two test points added to `test_utility_funs.py`, all the nodes that support constant-folding in onnx graphs are being tested.

Commit:
Added unit tests for all supported nodes.

spandantiwari commented:
@houseroad - as per your suggestion, I have now added unit tests for each individual node type supported for constant folding (`Concat`, `Unsqueeze`, `Transpose`, `Slice`). In addition,  the end-to-end tests for larger models (e.g. using LSTM) are also there. As you suggested, for testing purpose only, I have added a flag `_disable_torch_constant_prop` to `_model_to_graph()` that I use in the test points. Also, the new test points required support for constant folding `onnx::Cast` nodes also, which is also now added.

@houseroad and @suo - could you please review and see if this is ready for merging. Thanks!

Commit:
Explicitly adding opset version in unit test invoking _model_to_graph().

houseroad reviewed (dismissed):
Looks good to me now. Thanks for adding this useful pass to onnx exporter.

Nit: git clang-format on your changes?

@suo could you please take another look? :-)

houseroad commented:
@spandantiwari the tests added in test_util_func.py are failing, could you please fix them?

Commit:
Fixing bug in new unit tests.

suo commented:
The changes look good to me overall, we just need to figure out why `caffe2_onnx_py2_gcc5_ubuntu16_04_test` is failing. @zrphercule can you help debug it?

Commit:
Fixing issue with setting _export_onnx_opset_version in symbolic.py.

Commit:
Adding messages with assert in unit tests to root cause CI failure.

spandantiwari commented:
Wasn't able to repro the test failure on my side. But based on a hunch, I updated the PR with a fix related to `_export_onnx_opset_version`. Also added some debug info to the failing asserts, which I plan to remove once all the tests start to pass.

spandantiwari commented:
@houseroad, @suo - Fingers crossed, but the failing platform from previous runs is passing now. The issue seems to have been in setting the opset version using `torch.onnx.symbolic._set_opset_version`, which I have updated. 

I will wait for the all the runs to finish, and if all green I will update the PR to clean up the debugging instrumentation I added in last commit.

houseroad commented:
@spandantiwari the tests are passing now :-)

Commit:
Cleaning debugging messages.

spandantiwari commented:
> 

> @spandantiwari the tests are passing now :-)

@houseroad - Yes. And I have updated the PR to clean up any debug-related messages too.

houseroad commented:
lint please

Commit:
Fixing lint.

houseroad reviewed (approved):
Looks good to me.

houseroad commented:
@suo, friendly ping for another review :-)

suo commented:
This looks good to merge

spandantiwari commented:
@houseroad @suo - thank you for your review and all the work on this. Much appreciated.

spandantiwari commented on test/onnx/test_utility_funs.py:
```diff
@@ -25,6 +27,115 @@ def forward(self, x):
+            assert node.kind() != "onnx::Concat"
+            assert node.kind() != "onnx::Cast"
+            assert node.kind() != "onnx::Constant"
+        assert len(list(graph.nodes())) == 1
```
@ezyang - I would love to be able to improve this test point (or constant folding for that matter), but I am not sure I fully understand the question. 

First, as I asked in https://github.com/pytorch/pytorch/issues/28834 also (I saw that issue before I saw your comment here), I am not sure I understand what is meant by initialization graph. 

The goal of this test is simple. I have a graph where I have two constants feeding into a `Concat` node, followed by an `Add` op. Regarding the `Cast` op, I recall that they were not intentionally part of this test, but appeared in the graph when we set `_disable_torch_constant_prop=True`. However, since `Cast` is supported in constant folding they get folded too. Anyway, this test was supposed to test that the `Concat` node is constant folded (together with `Cast` op). The new folded constant after collapsing the `onnx::Constant` and the `onnx::Concat` node shows up as onnx initializer and not `onnx::Constant` in the final ONNX graph. This is why we check that there are no `onnx::Constant`, cast, or concat nodes. 

As an aside - looking back at this I feel this test would be better if I use the input `x` to the `forward()`  method in the model. I think it will make things clearer. I will try to update the test.

spandantiwari replied:
> what I expect to happen here is to get a graph with a single node that returns a constant. 

Forgot to mention - we do not support `Add` in constant folding yet, but if we did you are right in that the Add op should have been folded too. You would be left with a single intializer in the final ONNX graph.
